test/e2e/systeminternaltls/system_internal_tls_test.go

```diff
@@ -113,45 +107,162 @@ func TestSystemInternalTLS(t *testing.T) {
 		t.Fatalf("No pods detected for test app: %v", err)
 	}
 	helloWorldPod := pods.Items[0]
-	req = clients.KubeClient.CoreV1().Pods(helloWorldPod.Namespace).GetLogs(helloWorldPod.Name, &corev1.PodLogOptions{Container: "queue-proxy"})
-	queueTLSCount, err := scanPodLogs(req, matchTLSLog)
 
-	if err != nil {
-		t.Fatalf("Failed checking queue-proxy logs: %s", err)
-	} else if queueTLSCount == 0 {
-		t.Fatal("TLS not used on requests to queue-proxy")
+	if err := e2e.WaitForLog(t, clients, helloWorldPod.Namespace, helloWorldPod.Name, "queue-proxy", matchTLSLog, numMatches); err != nil {
+		t.Fatal("TLS not used on requests to queue-proxy:", err)
 	}
 }
 
-func scanPodLogs(req *rest.Request, matcher func(string) bool) (matchCount int, err error) {
+// TestTLSCertificateRotation tests certificate rotation and automatic reloading of certs.
+func TestTLSCertificateRotation(t *testing.T) {
+	if !test.ServingFlags.EnableAlphaFeatures {
+		t.Skip("Alpha features not enabled")
+	}
+
+	if !(strings.Contains(test.ServingFlags.IngressClass, "kourier") ||
+		strings.Contains(test.ServingFlags.IngressClass, "istio")) {
+		t.Skip("Skip this test for non-kourier or non-istio ingress.")
+	}
+
+	t.Parallel()
+
+	clients := test.Setup(t)
 
-	podLogs, err := req.Stream(context.Background())
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   test.HelloWorld,
+	}
+
+	t.Log("Creating Service:", names.Service)
+	resources1, err := v1test.CreateServiceReady(t, clients, &names,
+		rtesting.WithConfigAnnotations(map[string]string{
+			// Make sure we don't scale to zero during the as we're waiting for logs.
+			autoscaling.MinScaleAnnotationKey: "1",
+		}))
 	if err != nil {
-		err = fmt.Errorf("failed to stream activator logs: %w", err)
-		return
+		t.Fatalf("Failed to create Service: %v: %v", names.Service, err)
 	}
 
-	buf := new(bytes.Buffer)
-	_, err = io.Copy(buf, podLogs)
-	podLogs.Close()
+	t.Log("Checking Endpoint state")
+	url := resources1.Route.Status.URL.URL()
+	checkEndpointState(t, clients, url)
+
+	secret, err := e2e.GetCASecret(clients)
 	if err != nil {
-		err = fmt.Errorf("failed to read activator logs from buffer: %w", err)
-		return
+		t.Fatal(err)
+	}
+	if err := clients.KubeClient.CoreV1().Secrets(secret.Namespace).Delete(context.Background(), secret.Name, v1.DeleteOptions{}); err != nil {
```

Comment:
Hm, I assume this could have a side-effect. Not sure how cert-manager reacts to that. How about creating a new unrelated cert-manager certificate and using that certificate one to update the existing one? I probably would not just delete it.

Reply:
Another approach would be creating a new ClusterIssuer and re-configure knative ConfigMap `config-certmanager` to reference this new issuer. But that change Knative configuration during the test which is uncommon in this repository, and it would also not work when these configmap are managed by the knative operator (changing the config manually would not take effect as the operator would revert it).